src/prof/src/lib.rs

```diff
@@ -35,6 +35,181 @@ pub struct StackProfile {
     stacks: Vec<(WeightedStack, Option<usize>)>,
 }
 
+/// Gets the GNU build IDs for all loaded images, including the main
+/// program binary as well as all dynamically loaded libraries.
+/// Intended to be useful for profilers, who can use the supplied IDs
+/// to symbolicate stack traces offline.
+///
+/// Uses `dl_iterate_phdr` to walk the program headers of all images,
+/// and iterates over them looking for note
+/// segments. Then searches the discovered note segments for a note of type
+/// `NT_GNU_BUILD_ID` (aka "3") and name "GNU\0".
+///
+/// SAFETY: This function is written in a hilariously unsafe way: it involves
+/// following pointers to random parts of memory, and then assuming
+/// that particular structures can be found there.
+/// However, it was written by carefully reading `man dl_iterate_phdr`
+/// and `man elf`, and is thus intended to be relatively safe for callers to use.
+/// Assuming I haven't written any bugs (and that the documentation is correct),
+/// the only known safety requirements are:
+///
+/// (1) It must not be called multiple times concurrently, as `dl_iterate_phdr`
+/// is not documented as being thread-safe.
+/// (2) The running binary must be in ELF format and running on Linux.
+#[cfg(not(target_os = "macos"))]
+pub unsafe fn all_build_ids(
+) -> Result<std::collections::HashMap<std::path::PathBuf, Vec<u8>>, anyhow::Error> {
+    // local imports to avoid polluting the namespace for macOS builds
+    use std::collections::hash_map::Entry;
+    use std::collections::HashMap;
+    use std::ffi::{c_int, CStr, OsStr};
+    use std::os::unix::ffi::OsStrExt;
+    use std::path::{Path, PathBuf};
+
+    use mz_ore::cast::CastFrom;
+
+    use anyhow::Context;
+    use libc::{dl_iterate_phdr, dl_phdr_info, size_t, Elf64_Word, PT_NOTE};
+
+    struct CallbackState {
+        map: HashMap<PathBuf, Vec<u8>>,
+        is_first: bool,
+        fatal_error: Option<anyhow::Error>,
+    }
+    extern "C" fn iterate_cb(info: *mut dl_phdr_info, _size: size_t, data: *mut c_void) -> c_int {
+        // SAFETY: `data` is a pointer to a `CallbackState`, and no mutable reference
+        // aliases with it in Rust. Furthermore, `dl_iterate_phdr` doesn't do anything
+        // with `data` other than pass it to this callback, so nothing will be mutating
+        // the object it points to while we're inside here.
+        let state: &mut CallbackState = unsafe {
+            (data as *mut CallbackState)
+                .as_mut()
+                .expect("`data` cannot be null")
+        };
+        // SAFETY: similarly, `dl_iterate_phdr` isn't mutating `info`
+        // while we're here.
+        let info = unsafe { info.as_ref() }.expect("`dl_phdr_info` cannot be null");
+        let fname = if state.is_first {
+            // From `man dl_iterate_phdr`:
+            // "The first object visited by callback is the main program.  For the main
+            // program, the dlpi_name field will be an empty string."
```

Comment:
Trivia: musl will set dlpi_name to "/proc/self/exe" for the first object in the static case.

Reply:
Do you mean it sets it to the path /proc/self/exe points to, or literally to the string "/proc/self/exe" ?

Reply:
Literally this string (presumably leaving the caller to realpath/readlink this only if it needs to).